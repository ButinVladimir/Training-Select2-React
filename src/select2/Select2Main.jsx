import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import Select2Container from './Select2Container';
import ListItemValue from './helpers/list-item-value';
import sortFunc from './helpers/sort-func';

export default class Select2Main extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showPopupMenu: false,
      searchQuery: '',
      foundItems: [],
      selectedItems: [],
    };

    this.onClickOutsideContainer = this.onClickOutsideContainer.bind(this);
    this.onClickSelectedText = this.onClickSelectedText.bind(this);
    this.onSearchQueryChange = this.onSearchQueryChange.bind(this);
    this.onChangeItemSelection = this.onChangeItemSelection.bind(this);

    this.containerRef = createRef();
    this.throttleTimeout = null;
    this.queryId = 0;
  }

  componentDidMount() {
    window.addEventListener('click', this.onClickOutsideContainer);
  }

  componentWillUnmount() {
    if (this.throttleTimeout) {
      clearTimeout(this.throttleTimeout);
    }
    window.removeEventListener('click', this.onClickOutsideContainer);
  }

  onClickSelectedText() {
    this.setState({
      showPopupMenu: true,
      searchQuery: '',
      foundItems: this.restoreSelectedItems(),
    });
  }

  async onSearchQueryChange(event) {
    const searchQuery = event.target.value;

    if (this.throttleTimeout) {
      clearTimeout(this.throttleTimeout);
    }

    const newQueryId = Date.now();
    this.queryId = newQueryId;

    if (searchQuery) {
      this.setState({ searchQuery });

      this.throttleTimeout = setTimeout(async () => {
        if (this.queryId === newQueryId) {
          this.setState({ foundItems: await this.getAndSortData(searchQuery) });
        }
      }, 1000);
    } else {
      this.setState({ searchQuery, foundItems: this.restoreSelectedItems() });
    }
  }

  onChangeItemSelection(value, name) {
    const { performSelection, onSelect } = this.props;
    const { selectedItems, foundItems } = this.state;

    const {
      newSelectedItems,
      newFoundItems,
    } = performSelection(value, name, selectedItems, foundItems);
    onSelect(newSelectedItems.map(nsv => nsv.value));

    this.setState({
      selectedItems: newSelectedItems,
      foundItems: newFoundItems,
    });
  }

  onClickOutsideContainer(event) {
    const { showPopupMenu } = this.state;
    if (showPopupMenu
        && this.containerRef.current
        && !this.containerRef.current.contains(event.target)) {
      this.setState({ showPopupMenu: false });
    }
  }

  async getAndSortData(searchQuery) {
    const { getData } = this.props;
    const { selectedItems } = this.state;

    return (await getData(searchQuery))
      .map(foundItem => new ListItemValue(
        foundItem.name,
        foundItem.value,
        selectedItems.some(sv => sv.value === foundItem.value),
      ))
      .sort(sortFunc);
  }

  restoreSelectedItems() {
    const { selectedItems } = this.state;

    return selectedItems.concat().sort(sortFunc);
  }

  render() {
    const {
      prepareSelectedText,
      SelectedText,
      PopupMenu,
      InputTextField,
      ItemsList,
      ListItem,
    } = this.props;
    const {
      showPopupMenu,
      searchQuery,
      foundItems,
      selectedItems,
    } = this.state;

    const selectedText = prepareSelectedText(selectedItems);

    return (
      <Select2Container ref={this.containerRef}>
        <SelectedText value={selectedText} onClick={this.onClickSelectedText} />
        {showPopupMenu
          && (
          <PopupMenu
            inputTextField={(
              <InputTextField
                searchQuery={searchQuery}
                onChange={this.onSearchQueryChange}
              />
            )}

            itemsList={(
              <ItemsList
                ListItem={ListItem}
                items={foundItems}
                onChange={this.onChangeItemSelection}
              />
            )}
          />)}
      </Select2Container>
    );
  }
}

Select2Main.propTypes = {
  onSelect: PropTypes.func.isRequired,
  getData: PropTypes.func.isRequired,
  prepareSelectedText: PropTypes.func.isRequired,
  performSelection: PropTypes.func.isRequired,
  SelectedText: PropTypes.func.isRequired,
  PopupMenu: PropTypes.func.isRequired,
  InputTextField: PropTypes.func.isRequired,
  ItemsList: PropTypes.func.isRequired,
  ListItem: PropTypes.func.isRequired,
};
