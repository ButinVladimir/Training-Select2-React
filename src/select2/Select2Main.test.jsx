import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createSerializer } from 'enzyme-to-json';
import Select2Main from './Select2Main';
import InputTextField from './components/input-text-field/DefaultInputTextField';
import ItemsList from './components/items-list/DefaultItemsList';
import ListItem from './components/items-list/DefaultListItemSingle';
import PopupMenu from './components/popup-menu/DefaultPopupMenu';
import SelectedText from './components/selected-text/DefaultSelectedText';
import getData from '../get-data-local';
import performSelection from './helpers/perform-selection-single';
import prepareSelectedText from './helpers/default-prepare-selected-text';
import ListItemValue from './helpers/list-item-value';

describe('Select2Main', () => {
  beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
    expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));
  });

  const mountWrapper = onSelect => mount(
    <Select2Main
      onSelect={onSelect}
      getData={getData}
      prepareSelectedText={prepareSelectedText}
      performSelection={performSelection}
      SelectedText={SelectedText}
      PopupMenu={PopupMenu}
      InputTextField={InputTextField}
      ItemsList={ItemsList}
      ListItem={ListItem}
    />,
  );

  test('renders without crashing', () => {
    const wrapper = mountWrapper(() => {});

    expect(wrapper).toMatchSnapshot();
  });

  test('shows popup properly when no items are selected', () => {
    const wrapper = mountWrapper(() => {});

    wrapper.find(SelectedText).simulate('click');
    const state = wrapper.state();
    expect(state.showPopupMenu).toEqual(true);
    expect(state.searchQuery).toEqual('');
    expect(state.foundItems).toMatchObject([]);

    expect(wrapper).toMatchSnapshot();
  });

  test('shows popup properly when some items are selected', () => {
    const wrapper = mountWrapper(() => {});

    const selectedItems = [new ListItemValue('name', 'value', true)];
    wrapper.setState({ selectedItems });
    wrapper.find(SelectedText).simulate('click');
    const state = wrapper.state();
    expect(state.showPopupMenu).toEqual(true);
    expect(state.searchQuery).toEqual('');
    expect(state.foundItems).toMatchObject(selectedItems);

    expect(wrapper).toMatchSnapshot();
  });

  test('fires timer after text in input field has been changed', () => {
    jest.useFakeTimers();
    const wrapper = mountWrapper(() => {});

    wrapper.find(SelectedText).simulate('click');
    const input = wrapper.find(InputTextField).find('input');
    input.instance().value = 'a';
    input.simulate('change');

    expect(clearTimeout).toHaveBeenCalledTimes(0);
    expect(setTimeout).toHaveBeenCalledTimes(1);
    jest.runAllTimers();

    const state = wrapper.state();
    expect(state.showPopupMenu).toEqual(true);
    expect(state.searchQuery).toEqual('a');
    expect(state.foundItems).toMatchObject([]);

    expect(wrapper).toMatchSnapshot();
  });

  test('clears and fires timer once after text in input field has been changed twice', () => {
    jest.useFakeTimers();
    const wrapper = mountWrapper(() => {});

    wrapper.find(SelectedText).simulate('click');
    const input = wrapper.find(InputTextField).find('input');
    input.instance().value = 'a';
    input.simulate('change');
    input.instance().value = 'alpha';
    input.simulate('change');

    expect(clearTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenCalledTimes(2);
    jest.runAllTimers();

    const state = wrapper.state();
    expect(state.showPopupMenu).toEqual(true);
    expect(state.searchQuery).toEqual('alpha');
    expect(state.foundItems).toMatchObject([]);

    expect(wrapper).toMatchSnapshot();
  });

  test('fires timer after text in input field has been cleared', () => {
    jest.useFakeTimers();
    const wrapper = mountWrapper(() => {});
    const selectedItems = [new ListItemValue('name', 'value', true)];

    wrapper.setState({ selectedItems });
    wrapper.find(SelectedText).simulate('click');
    const input = wrapper.find(InputTextField).find('input');
    input.instance().value = 'a';
    input.simulate('change');
    input.instance().value = '';
    input.simulate('change');

    expect(clearTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenCalledTimes(1);
    jest.runAllTimers();

    const state = wrapper.state();
    expect(state.showPopupMenu).toEqual(true);
    expect(state.searchQuery).toEqual('');
    expect(state.foundItems).toMatchObject(selectedItems);
    expect(state.selectedItems).toMatchObject(selectedItems);

    expect(wrapper).toMatchSnapshot();
  });

  test('clears timer after component is unmounted', () => {
    jest.useFakeTimers();
    const wrapper = mountWrapper(() => {});
    const selectedItems = [new ListItemValue('name', 'value', true)];

    wrapper.setState({ selectedItems });
    wrapper.find(SelectedText).simulate('click');
    const input = wrapper.find(InputTextField).find('input');
    input.instance().value = 'a';
    input.simulate('change');

    expect(clearTimeout).toHaveBeenCalledTimes(0);
    expect(setTimeout).toHaveBeenCalledTimes(1);
    wrapper.unmount();
    expect(clearTimeout).toHaveBeenCalledTimes(1);
  });

  test('gets and sorts data correctly when nothing is selected', async () => {
    const wrapper = mountWrapper(() => {});

    const foundItems = await wrapper.instance().getAndSortData('alpha');
    expect(foundItems).toMatchObject([
      new ListItemValue('Alpha', 1, false),
      new ListItemValue('Alpha 1', 11, false),
      new ListItemValue('Alpha 2', 21, false),
    ]);
  });

  test('gets and sorts data correctly when something is selected', async () => {
    const wrapper = mountWrapper(() => {});

    wrapper.setState({ selectedItems: [new ListItemValue('Alpha 1', 11, true)] });

    const foundItems = await wrapper.instance().getAndSortData('alpha');
    expect(foundItems).toMatchObject([
      new ListItemValue('Alpha 1', 11, true),
      new ListItemValue('Alpha', 1, false),
      new ListItemValue('Alpha 2', 21, false),
    ]);
  });

  test('changes selected item correctly', async () => {
    const mockFn = jest.fn();
    const wrapper = mountWrapper(mockFn);

    const foundItems = [
      new ListItemValue('Alpha 1', 11, true),
      new ListItemValue('Alpha', 1, false),
      new ListItemValue('Alpha 2', 21, false),
    ];
    const selectedItems = [
      new ListItemValue('Alpha 1', 11, true),
    ];
    wrapper.setState({ selectedItems, foundItems });
    wrapper.instance().onChangeItemSelection(21, 'Alpha 2');

    const state = wrapper.state();

    expect(state.foundItems).toEqual([
      new ListItemValue('Alpha 1', 11, false),
      new ListItemValue('Alpha', 1, false),
      new ListItemValue('Alpha 2', 21, true),
    ]);
    expect(state.selectedItems).toEqual([new ListItemValue('Alpha 2', 21, true)]);
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledWith([21]);
    expect(wrapper).toMatchSnapshot();
  });
});
