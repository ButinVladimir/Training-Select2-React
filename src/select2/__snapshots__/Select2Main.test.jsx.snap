// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select2Main changes selected item correctly 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="Alpha 1"
    />
  </div>
</ForwardRef>
`;

exports[`Select2Main clears and fires timer once after text in input field has been changed twice 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="Click to select"
    />
    <div
      className="select2-popup-menu"
    >
      <div
        className="input-text-container"
      >
        <input
          className="select2-input-text-field"
          onChange={[Function]}
          type="text"
          value="alpha"
        />
      </div>
      <div>
        <div
          className="items-not-found"
        >
          No matching items found
        </div>
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Select2Main fires timer after text in input field has been changed 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="Click to select"
    />
    <div
      className="select2-popup-menu"
    >
      <div
        className="input-text-container"
      >
        <input
          className="select2-input-text-field"
          onChange={[Function]}
          type="text"
          value="a"
        />
      </div>
      <div>
        <div
          className="items-not-found"
        >
          No matching items found
        </div>
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Select2Main fires timer after text in input field has been cleared 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="name"
    />
    <div
      className="select2-popup-menu"
    >
      <div
        className="input-text-container"
      >
        <input
          className="select2-input-text-field"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div>
        <ul
          className="select2-item-list"
        >
          <li
            className="checked"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              name
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Select2Main renders without crashing 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="Click to select"
    />
  </div>
</ForwardRef>
`;

exports[`Select2Main shows popup properly when no items are selected 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="Click to select"
    />
    <div
      className="select2-popup-menu"
    >
      <div
        className="input-text-container"
      >
        <input
          className="select2-input-text-field"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div>
        <div
          className="items-not-found"
        >
          No matching items found
        </div>
      </div>
    </div>
  </div>
</ForwardRef>
`;

exports[`Select2Main shows popup properly when some items are selected 1`] = `
<ForwardRef>
  <div
    className="select2-container"
  >
    <input
      className="select2-selected-text"
      onClick={[Function]}
      readOnly={true}
      type="text"
      value="name"
    />
    <div
      className="select2-popup-menu"
    >
      <div
        className="input-text-container"
      >
        <input
          className="select2-input-text-field"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div>
        <ul
          className="select2-item-list"
        >
          <li
            className="checked"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              name
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</ForwardRef>
`;
